torch/csrc/distributed/rpc/rpc_agent.h

```diff
@@ -63,7 +53,9 @@ class RpcAgent {
   // If ``message.isRequest()`` is true, the ``FutureMessage`` will be completed
   // when the response arrives. For other message types, the Future should be
   // ignored by the caller.
-  std::shared_ptr<FutureMessage> send(const WorkerId& to, Message&& message);
+  virtual std::shared_ptr<FutureMessage> send(
+      const WorkerId& to,
+      Message&& message) = 0;
```

Comment:
(1) Since you added an implementation of send in rpc_agent.cpp this shouldn't be virtual nor pure virtual and only sendImpl should be pure virtual. (2) I am not sure what is the boundary for what is rpc agent should be responsible for, with this diff we are making the agent aware of autograd and in a way breaking the single responsibility principle. I am fine with this if we are doing this to move fast, we probably should add a comment stating that we should re-look at it.